Userland/Applications/Maps/MapWidget.cpp

```diff
@@ -0,0 +1,295 @@
+    : m_tile_layer_url(options.tile_layer_url)
+    , m_center(options.center)
+    , m_zoom(options.zoom)
+    , m_scale_enabled(options.scale_enabled)
+    , m_scale_max_width(options.scale_max_width)
+    , m_attribution_enabled(options.attribution_enabled)
+    , m_attribution_text(options.attribution_text)
+    , m_attribution_url(options.attribution_url)
+{
+    m_request_client = Protocol::RequestClient::try_create().release_value_but_fixme_should_propagate_errors();
+}
+
+void MapWidget::mousedown_event(GUI::MouseEvent& event)
+{
+    if (m_connection_failed)
+        return;
+
+    if (event.button() == GUI::MouseButton::Primary) {
+        // Ignore attribution click
+        if (m_attribution_enabled && static_cast<float>(event.x()) > width() - m_attribution_width && static_cast<float>(event.y()) > height() - m_attribution_height) {
+            return;
+        }
+
+        // Start map tiles dragging
+        m_dragging = true;
+        m_last_mouse_x = event.x();
+        m_last_mouse_y = event.y();
+        set_override_cursor(Gfx::StandardCursor::Drag);
+    }
+}
+
+void MapWidget::mousemove_event(GUI::MouseEvent& event)
+{
+    if (m_connection_failed)
+        return;
+
+    if (m_dragging) {
+        // Adjust map center by mouse delta
+        double delta_x = event.x() - m_last_mouse_x;
+        double delta_y = event.y() - m_last_mouse_y;
+        set_center({ tile_y_to_latitude(latitude_to_tile_y(m_center.latitude, m_zoom) - delta_y / TILE_SIZE, m_zoom),
+            tile_x_to_longitude(longitude_to_tile_x(m_center.longitude, m_zoom) - delta_x / TILE_SIZE, m_zoom) });
+        m_last_mouse_x = event.x();
+        m_last_mouse_y = event.y();
+        return;
+    }
+
+    // Handle attribution hover
+    if (m_attribution_enabled) {
+        if (static_cast<float>(event.x()) > width() - m_attribution_width && static_cast<float>(event.y()) > height() - m_attribution_height) {
+            set_override_cursor(Gfx::StandardCursor::Hand);
+        } else {
+            set_override_cursor(Gfx::StandardCursor::Arrow);
+        }
+    }
+}
+
+void MapWidget::mouseup_event(GUI::MouseEvent& event)
+{
+    if (m_connection_failed)
+        return;
+
+    // Stop map tiles dragging
+    if (m_dragging) {
+        m_dragging = false;
+        set_override_cursor(Gfx::StandardCursor::Arrow);
+        return;
+    }
+
+    if (event.button() == GUI::MouseButton::Primary) {
+        // Handle attribution click
+        if (m_attribution_enabled && static_cast<float>(event.x()) > width() - m_attribution_width && static_cast<float>(event.y()) > height() - m_attribution_height) {
+            Desktop::Launcher::open(m_attribution_url);
+            return;
+        }
+    }
+}
+
+void MapWidget::mousewheel_event(GUI::MouseEvent& event)
+{
+    if (m_connection_failed)
+        return;
+
+    if (event.wheel_delta_y() > 0) {
+        set_zoom(m_zoom - 1);
+    } else {
+        set_zoom(m_zoom + 1);
+    }
```

Comment:
I'd expect the mouse wheel zoom to actually zoom in to your mouse cursor (i.e. by panning the map to the cursor first)

Reply:
You are right the code first only zoomed the center of the map. I have added the code that adjusts the zoom by the current mouse position.